### PR TITLE
Allow 'github' shorthand for extra-deps (fixes #3873)

### LINKED ISSUE #3873: Allow github shorthand for extra-deps
It's quite common to include an `extra-dep` from Github. The common way to do it, via a `git: https://github.com/...` URL, is actually _not_ the best. It's better to instead use a link to the archive, since this doesn't require downloading the entire Git history, and results in a faster transport. For example, see this commit:

https://github.com/commercialhaskell/stack/pull/3831/commits/638efeaef17a02cb00201bf85d9ef557999ae83f

I'd like to recommend that we add a convenience syntax that does the right thing, namely, replacing:

```
- https://github.com/commercialhaskell/rio/archive/09654f9fcbdcd96d0f5102796b32fdac5da7260e.tar.gz
```

with

```
- github: commercialhaskell/rio
  commit: 09654f9fcbdcd96d0f5102796b32fdac5da7260e
```

Or perhaps even:

```
- github: commercialhaskell/rio@09654f9fcbdcd96d0f5102796b32fdac5da7260e
```

And have this generate the archive link described above.

### PR DESCRIPTION
**Pull request by mitchellwrosen**
First, sorry to @lhcopetti for stealing this one - I neglected to notice you were interested in doing this.

This patch adds support for this syntax of `extra-deps`:

```
- github: commercialhaskell/rio
  commit: 09654f9fcbdcd96d0f5102796b32fdac5da7260e
```

with an additional optional `subdirs` field. The above would resolve to this archive:

```
https://github.com/commercialhaskell/rio/archive/09654f9fcbdcd96d0f5102796b32fdac5da7260e.tar.gz
```

Please let me know if I should tweak anything, including adding more tests. Thanks!

### CODE REVIEW

File: src/Data/Aeson/Extended.hs [hunk 1]
```diff
@@ -150,7 +150,7 @@ instance IsString WarningParserMonoid where
 
 -- Parsed JSON value with its warnings
 data WithJSONWarnings a = WithJSONWarnings a [JSONWarning]
-    deriving Generic
+    deriving (Eq, Generic, Show)
```
Comment (snoyberg) [inline src/Data/Aeson/Extended.hs]: Is this change necessary for the PR, or just to help with debugging?

File: src/Stack/Types/BuildPlan.hs [hunk 2]
```diff
@@ -269,6 +270,26 @@ instance subdirs ~ Subdirs => FromJSON (WithJSONWarnings (PackageLocation subdir
             , archiveHash = msha'
             }
 
+        github = withObjectWarnings "PLArchive" $ \o -> do
```
Comment (snoyberg) [inline src/Stack/Types/BuildPlan.hs]: It should probably say `PLArchive:github` or something like that instead, this is just used for error messages, and it would be useful to know what part of the code failed the parsing.

File: src/test/Stack/Types/BuildPlanSpec.hs [hunk 3]
```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Stack.Types.BuildPlanSpec where
```
Comment (snoyberg) [inline src/test/Stack/Types/BuildPlanSpec.hs]: Nice! 👍

File: src/Data/Aeson/Extended.hs (hunk 1 above)
Comment (mitchellwrosen) [reply src/Data/Aeson/Extended.hs]: `Eq` and `Show` are required for test code like:

```haskell
decode' contents `shouldBe` Just (WithJSONWarnings expected [])
```

File: src/Stack/Types/BuildPlan.hs (hunk 2 above)
Comment (mitchellwrosen) [reply src/Stack/Types/BuildPlan.hs]: Done

**Comment by mitchellwrosen** [comment]
Of course!
